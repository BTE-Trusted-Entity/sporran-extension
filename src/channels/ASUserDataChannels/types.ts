import { ICredential, KiltAddress } from '@kiltprotocol/types';
import { HexString } from '@polkadot/util/types';

import { DAppName } from '../AccessChannels/DAppName';
import { Origin } from '../AccessChannels/Origin';

export type ASUserDataInput = DAppName & {
  submitter: KiltAddress;
};

export type ASUserDataOriginInput = ASUserDataInput & Origin;

export interface ASUserDataOutput {
  createDidExtrinsic: HexString;
  credential: ICredential;
}
