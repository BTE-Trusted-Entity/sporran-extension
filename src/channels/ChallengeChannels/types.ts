import { DidResourceUri } from '@kiltprotocol/sdk-js';

export interface ChallengeInput {
  dAppEncryptionKeyId: DidResourceUri;
  challenge: string;
}

export interface ChallengeOutput {
  encryptionKeyId: DidResourceUri;
  encryptedChallenge: string;
  nonce: string;
}
