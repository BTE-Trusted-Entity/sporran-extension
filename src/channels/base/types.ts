export interface ErrorFirstCallback<Output> {
  (error: null, output: Output): void;

  (error: Error): void;
}
