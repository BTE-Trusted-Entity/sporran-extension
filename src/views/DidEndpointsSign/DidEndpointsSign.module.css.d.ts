export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const didLine: string;
export const fee: string;
export const details: string;
export const detail: string;
export const fullWidthDetail: string;
export const detailName: string;
export const detailValue: string;
export const buttonsLine: string;
export const cancel: string;
export const submit: string;
export const errorTooltip: string;
