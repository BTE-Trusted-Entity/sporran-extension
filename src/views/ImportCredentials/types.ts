import { KiltAddress } from '@kiltprotocol/sdk-js';

export interface Import {
  fileName: string;
}

export interface SuccessfulImport extends Import {
  identityAddress: KiltAddress;
}

export interface FailedImport extends Import {
  error: 'invalid' | 'orphaned' | string;
}
