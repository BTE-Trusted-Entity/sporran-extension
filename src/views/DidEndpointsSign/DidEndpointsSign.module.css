.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.didLine {
  composes: didLine from '../../components/CopyValue/CopyValue.module.css';
  margin-top: 1rem;
}

.fee {
  composes: smallText from '../../components/Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 0.5rem 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.details {
  composes: details from '../../components/Card/Card.module.css';
}

.detail {
  composes: detail from '../../components/Card/Card.module.css';
}

.fullWidthDetail {
  composes: fullWidthDetail from '../../components/Card/Card.module.css';
}

.detailName {
  composes: detailName from '../../components/Card/Card.module.css';
}

.detailValue {
  composes: detailValue from '../../components/Card/Card.module.css';
  font-size: 0.875rem;
  max-width: 100%;
  overflow-x: auto;
  white-space: nowrap;
  scrollbar-width: none;
}

.detailValue::-webkit-scrollbar {
  display: none;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin-top: 0.5rem;
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
