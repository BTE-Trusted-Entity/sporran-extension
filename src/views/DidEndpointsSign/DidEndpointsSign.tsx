import { FormEvent, useCallback } from 'react';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import {
  ConfigService,
  Did,
  DidServiceEndpoint,
  DidUri,
  KiltKeyringPair,
  SignExtrinsicCallback,
} from '@kiltprotocol/sdk-js';

import * as styles from './DidEndpointsSign.module.css';

import { Identity } from '../../utilities/identities/types';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import {
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import { CopyValue } from '../../components/CopyValue/CopyValue';
import {
  getIdentityCryptoFromSeed,
  getIdentityDid,
} from '../../utilities/identities/identities';
import { useSubmitStates } from '../../utilities/useSubmitStates/useSubmitStates';
import { getFullDidDocument } from '../../utilities/did/did';
import { generatePath, paths } from '../paths';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { KiltAmount } from '../../components/KiltAmount/KiltAmount';
import { makeFakeIdentityCrypto } from '../../utilities/makeFakeIdentityCrypto/makeFakeIdentityCrypto';

type ActionType = 'add' | 'remove';

async function getDidAuthorizedExtrinsic(
  keypair: KiltKeyringPair,
  sign: SignExtrinsicCallback,
  did: DidUri,
  service: DidServiceEndpoint,
  type: ActionType,
) {
  const document = await getFullDidDocument(did);

  const api = ConfigService.get('api');
  const draft =
    type === 'add'
      ? api.tx.did.addServiceEndpoint(Did.serviceToChain(service))
      : api.tx.did.removeServiceEndpoint(Did.resourceIdToChain(service.id));

  return Did.authorizeTx(document.uri, draft, sign, keypair.address);
}

async function getFee(
  did: DidUri,
  service: DidServiceEndpoint,
  type: ActionType,
) {
  const { keypair, sign } = makeFakeIdentityCrypto();

  const authorized = await getDidAuthorizedExtrinsic(
    keypair,
    sign,
    did,
    service,
    type,
  );

  return (await authorized.paymentInfo(keypair)).partialFee;
}

interface Props {
  identity: Identity;
  type: ActionType;
  endpoint: DidServiceEndpoint;
}

export function DidEndpointsSign({
  identity,
  type,
  endpoint,
}: Props): JSX.Element {
  const t = browser.i18n.getMessage;
  const { address } = identity;
  const did = getIdentityDid(identity);

  const fee = useAsyncValue(getFee, [did, endpoint, type]);

  const passwordField = usePasswordField();
  const { submit, modalProps, submitting, unpaidCosts } = useSubmitStates();

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      const { keypair, seed } = await passwordField.get(event);
      const { sign } = await getIdentityCryptoFromSeed(seed);

      const authorized = await getDidAuthorizedExtrinsic(
        keypair,
        sign,
        did,
        endpoint,
        type,
      );

      await submit(keypair, authorized);
    },
    [passwordField, type, endpoint, did, submit],
  );

  const modalMessagesAdd = {
    pending: t('view_DidEndpointsSign_add_pending'),
    success: t('view_DidEndpointsSign_add_success'),
    error: t('view_DidEndpointsSign_add_error'),
  };
  const modalMessagesRemove = {
    pending: t('view_DidEndpointsSign_remove_pending'),
    success: t('view_DidEndpointsSign_remove_success'),
    error: t('view_DidEndpointsSign_remove_error'),
  };

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>
        {type === 'add'
          ? t('view_DidEndpointsSign_heading_add')
          : t('view_DidEndpointsSign_heading_remove')}
      </h1>
      <p className={styles.subline}>{t('view_DidEndpointsSign_subline')}</p>

      <IdentitySlide identity={identity} />

      <CopyValue value={did} label="DID" className={styles.didLine} />

      {fee && (
        <p className={styles.fee}>
          {t('view_DidEndpointsSign_fee')}
          <KiltAmount amount={fee} type="costs" smallDecimals />
        </p>
      )}

      <dl className={styles.details}>
        <div className={styles.fullWidthDetail}>
          <dt className={styles.detailName}>
            {t('view_DidEndpointsSign_url')}
          </dt>
          <dd className={styles.detailValue}>{endpoint.serviceEndpoint[0]}</dd>
        </div>
        <div className={styles.detail}>
          <dt className={styles.detailName}>
            {t('view_DidEndpointsSign_type')}
          </dt>
          <dd className={styles.detailValue}>{endpoint.type[0]}</dd>
        </div>
        <div className={styles.detail}>
          <dt className={styles.detailName}>{t('view_DidEndpointsSign_id')}</dt>
          <dd className={styles.detailValue}>
            {Did.resourceIdToChain(endpoint.id)}
          </dd>
        </div>
      </dl>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button type="submit" className={styles.submit} disabled={submitting}>
          {t('common_action_sign')}
        </button>
        <output
          className={styles.errorTooltip}
          hidden={!unpaidCosts || Boolean(modalProps)}
        >
          {t('view_DidEndpointsSign_insufficientFunds', [unpaidCosts])}
        </output>
      </p>

      {modalProps && (
        <TxStatusModal
          {...modalProps}
          identity={identity}
          messages={type === 'add' ? modalMessagesAdd : modalMessagesRemove}
          destination={generatePath(paths.identity.did.manage.endpoints.start, {
            address,
          })}
        />
      )}

      <LinkBack />
      <Stats />
    </form>
  );
}
