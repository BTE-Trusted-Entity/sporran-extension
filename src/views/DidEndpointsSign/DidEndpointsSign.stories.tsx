import { Meta } from '@storybook/react';
import { DidServiceEndpoint } from '@kiltprotocol/sdk-js';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { DidEndpointsSign } from './DidEndpointsSign';

export default {
  title: 'Views/DidEndpointsSign',
  component: DidEndpointsSign,
} as Meta;

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

const endpoint: DidServiceEndpoint = {
  serviceEndpoint: ['https://sporran.org/'],
  type: ['Some Type'],
  id: '#123456',
};

const longUrlEndpoint: DidServiceEndpoint = {
  ...endpoint,
  serviceEndpoint: [
    'https://www.this-is-a-super-long-url/which-just-keeps-going-and-going/to-test-the-overflow-behaviour-of-the-url-value',
  ],
};

export function Add(): JSX.Element {
  return (
    <DidEndpointsSign type="add" identity={identity} endpoint={endpoint} />
  );
}

export function Remove(): JSX.Element {
  return (
    <DidEndpointsSign type="remove" identity={identity} endpoint={endpoint} />
  );
}

export function LongUrl(): JSX.Element {
  return (
    <DidEndpointsSign
      type="add"
      identity={identity}
      endpoint={longUrlEndpoint}
    />
  );
}
