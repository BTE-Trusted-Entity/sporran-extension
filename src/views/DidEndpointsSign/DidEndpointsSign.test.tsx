import { DidServiceEndpoint } from '@kiltprotocol/types';
import { DidUtils } from '@kiltprotocol/did';
import { BalanceUtils } from '@kiltprotocol/core';

import { identitiesMock, render } from '../../testing/testing';
import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import '../../components/useCopyButton/useCopyButton.mock';

import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';

import { DidEndpointsSign } from './DidEndpointsSign';

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

const endpoint: DidServiceEndpoint = {
  urls: ['https://sporran.org/'],
  types: ['Some Type'],
  id: '123456',
};

jest.mocked(DidUtils.parseDidUri).mockReturnValue({
  identifier: '4pehddkhEanexVTTzWAtrrfo2R7xPnePpuiJLC7shQU894aY',
} as ReturnType<typeof DidUtils.parseDidUri>);

jest.mock('../../utilities/useAsyncValue/useAsyncValue');
jest.mocked(useAsyncValue).mockReturnValue(BalanceUtils.toFemtoKilt(0.01));

describe('DidEndpointsSign', () => {
  it('should match the snapshot when adding', async () => {
    const { container } = render(
      <DidEndpointsSign type="add" identity={identity} endpoint={endpoint} />,
    );

    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot when removing', async () => {
    const { container } = render(
      <DidEndpointsSign
        type="remove"
        identity={identity}
        endpoint={endpoint}
      />,
    );

    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
