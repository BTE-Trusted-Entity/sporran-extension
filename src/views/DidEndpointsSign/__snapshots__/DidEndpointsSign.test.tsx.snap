// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidEndpointsSign should match the snapshot when adding 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Add Service Endpoint
    </h1>
    <p
      class="subline"
    >
      Sign with your password
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <p
      class="fee"
    >
      Fee: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <dl
      class="details"
    >
      <div
        class="fullWidthDetail"
      >
        <dt
          class="detailName"
        >
          URL
        </dt>
        <dd
          class="detailValue"
        >
          https://sporran.org/
        </dd>
      </div>
      <div
        class="detail"
      >
        <dt
          class="detailName"
        >
          Type
        </dt>
        <dd
          class="detailValue"
        >
          Some Type
        </dd>
      </div>
      <div
        class="detail"
      >
        <dt
          class="detailName"
        >
          Service ID
        </dt>
        <dd
          class="detailValue"
        >
          123456
        </dd>
      </div>
    </dl>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least undefined KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;

exports[`DidEndpointsSign should match the snapshot when removing 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove Service Endpoint
    </h1>
    <p
      class="subline"
    >
      Sign with your password
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <p
      class="fee"
    >
      Fee: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <dl
      class="details"
    >
      <div
        class="fullWidthDetail"
      >
        <dt
          class="detailName"
        >
          URL
        </dt>
        <dd
          class="detailValue"
        >
          https://sporran.org/
        </dd>
      </div>
      <div
        class="detail"
      >
        <dt
          class="detailName"
        >
          Type
        </dt>
        <dd
          class="detailValue"
        >
          Some Type
        </dd>
      </div>
      <div
        class="detail"
      >
        <dt
          class="detailName"
        >
          Service ID
        </dt>
        <dd
          class="detailValue"
        >
          123456
        </dd>
      </div>
    </dl>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least undefined KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
